### ISSUE TITLE
Calculate contributions-views ratio

### ISSUE THREAD

**Issue by darenwelsh** [darenwelsh/UserJourney#33]
Use this query as an example of how to join the score and views query.

``` php
$query = "
SELECT score.day, (100 * `Admins score-views ratio` / `Admins views`) as ratio
FROM (
SELECT DATE(rev_timestamp) AS day, 
LEAST(100, COUNT(DISTINCT rev_page) + SQRT( COUNT(rev_id) - COUNT(DISTINCT rev_page) ) * 2 ) AS `Admins score-views ratio` 
FROM `revision` 
WHERE rev_user_text IN ( 'Ejmontal', 'Lwelsh', 'Swray', 'Balpert', 'Cmavridi', 'Ssjohns5', '' ) 
AND rev_timestamp >= '20150816000000' 
AND rev_timestamp < '20160422000000' 
GROUP BY day
) score
LEFT JOIN (
SELECT DATE(hit_timestamp) AS day, 
COUNT( * ) AS `Admins views` 
FROM `wiretap` 
WHERE user_name IN ( 'Ejmontal', 'Lwelsh', 'Swray', 'Balpert', 'Cmavridi', 'Ssjohns5', '' ) 
AND hit_timestamp >= '20150816000000' 
AND hit_timestamp < '20160422000000' 
GROUP BY day
) views
ON score.day = views.day;
";
```

**Comment by darenwelsh** [comment]
In order to facilitate this, create new function to generate SQL queries.

Then maybe do the following in PHP for the case of calulating a ratio of two subqueries:

``` php
$subquery1 = $this->generateQuery(`Admins score`);
$subquery2 = $this->generateQuery(`Admins views`);
$query = "
SELECT score.day, (100 * `Admins score` / `Admins views`) as ratio
FROM ( $subquery1 ) score
LEFT JOIN ( $subquery2 ) views
ON score.day = views.day;
";
```

``` php
function applyFormulaOnTwoQueries( ){
   $query = "
      SELECT score.day, (100 * `Admins score` / `Admins views`) as ratio
      FROM ( $subquery1 ) score
      LEFT JOIN ( $subquery2 ) views
      ON score.day = views.day;
   ";
}
```

**Comment by darenwelsh** [comment]
``` php
SELECT `Admins score`.day, (100 * `Admins score` / `Admins views`) as ratio
FROM (
SELECT 
    DATE(rev_timestamp) AS day, 
    LEAST(100, COUNT(DISTINCT rev_page) + SQRT( COUNT(rev_id) - COUNT(DISTINCT rev_page) ) * 2 ) AS `Admins score` 
FROM `revision` 
WHERE rev_user_text IN ( 'Ejmontal', 'Lwelsh', 'Swray', 'Balpert', 'Cmavridi', 'Ssjohns5', '' ) 
AND rev_timestamp >= '20150816000000' 
AND rev_timestamp < '20160422000000' 
GROUP BY day
) `Admins score`
LEFT JOIN (
SELECT 
    DATE(hit_timestamp) AS day, 
    COUNT( * ) AS `Admins views` 
FROM `wiretap` 
WHERE user_name IN ( 'Ejmontal', 'Lwelsh', 'Swray', 'Balpert', 'Cmavridi', 'Ssjohns5', '' ) 
AND hit_timestamp >= '20150816000000' 
AND hit_timestamp < '20160422000000' 
GROUP BY day
) `Admins views`
ON `Admins score`.day = `Admins views`.day
```

**Comment by darenwelsh** [comment]
``` php

$outputScore =
"INSERT INTO temp_union (day, `{$competitorName}`)

    SELECT
        DATE(rev_timestamp) AS day,
        LEAST({$wgUJscoreCeiling}, {$wgUJscoreDefinition} ) AS `{$competitorName}`
    FROM $revTable
    WHERE
        rev_user_text IN ( '{$person}', '{$competitorName}' )
    AND rev_timestamp >= '{$startDate}'
    AND rev_timestamp < '{$endDate}'
    GROUP BY day
";


$outputViews =
"INSERT INTO temp_union (day, `{$competitorName}`)

    SELECT
        DATE(hit_timestamp) AS day,
        COUNT(DISTINCT(CONCAT(user_name,'UNIQUESEPARATOR',page_id))) AS `{$competitorName}`
    FROM $wiretapTable
    WHERE
        user_name IN ( '{$person}', '{$competitorName}' )
    AND hit_timestamp >= '{$startDate}'
    AND hit_timestamp < '{$endDate}'
    GROUP BY day
";


function applyFormulaOnTwoQueries( $competitorName, $subquery1, $valueType1, $subquery2, $valueType2 ){
   $query = "SELECT 
                {$valueType1}.day, 
                (100 * {$valueType1}.`{$competitorName}` / {$valueType2}.`{$competitorName}`) as {$competitorName}
            FROM ( $subquery1 ) {$valueType1}
            LEFT JOIN ( $subquery2 ) {$valueType2}
            ON {$valueType1}.day = {$valueType2}.day;
   ";
}
```

**Comment by darenwelsh** [comment]
closed by 32c51d7
